### ISSUE TITLE
trtllm_fp4_block_scale_moe allocates FP32 topk_weights for DeepSeekV3 despite BF16 CUDA routing output

### ISSUE THREAD

**Issue by leejnau** [flashinfer-ai/flashinfer#3595]
## Summary

In `trtllm_fp4_block_scale_moe`, the Python wrapper allocates `topk_weights` using `routing_logits.dtype` when routing logits are provided. For FP32 routing logits, this creates an FP32  `topk_weights` buffer.

However, the C++ DeepSeekV3 routing path hard-codes the routing weight output dtype to BF16:

- Python allocation: `flashinfer/fused_moe/core.py`
- DeepSeekV3 routing output: `csrc/trtllm_fused_moe_runner.cu`

This creates a dtype/memory interpretation mismatch for:

```python
trtllm_fp4_block_scale_moe(
    routing_logits=fp32_logits,
    routing_method_type=RoutingMethodType.DeepSeekV3,
    ...
)
```

The issue is especially visible with `do_finalize=False`, where the returned expert_weights tensor can have FP32 metadata even though the routing kernel writes BF16 values.

## Potential Solution

For `routing_method_type == RoutingMethodType.DeepSeekV3`:

- Allocate internal topk_weights as `torch.bfloat16`, regardless of `routing_logits.dtype`.
- Validate user-provided unpacked `topk_weights` is BF16.
- Document that DeepSeekV3 expert_weights are BF16, including for `do_finalize=False`.

This matches the current C++ kernel behavior and the documented unpacked precomputed routing contract.

This is minimally invasive but changes the visible `dtype` of returned `expert_weights` for `do_finalize=False` from FP32 to BF16 in the FP32-logits + DeepSeekV3 case. That said, the current FP32 tensor appears to be incorrectly typed because the CUDA writer emits BF16 data.
